### PR TITLE
Verify Post-TC

### PR DESCRIPTION
**Pull request by zardyh**
Currently only that recursive definitions are well-formed.

```
test.ml[17:5 ..17:9]: error
  Invalid recursive right-hand side for variable x
  • Note: because the variable x is not under a function
  Arising from use of the declaration
   │ 
17 │ let x = x
   │ 
```

The check is very silly: None of the variables bound in the same binding
group can appear outside of a function. Implicit and explicit laziness
counts as "a function".

### CODE REVIEW

File: compiler/Main.hs [hunk 1]
```diff
@@ -72,14 +74,17 @@ compile (file:files) = runGen $ do
...
                 Right (prog, env') ->
-                  let (var, tys) = extractToplevels parsed
-                      (var', tys') = extractToplevels resolved
-                  in pure $ Right (tops ++ prog
-                                  , scope { RS.varScope = RS.insertN (RS.varScope scope) (zip var var')
-                                          , RS.tyScope  = RS.insertN (RS.tyScope scope)  (zip tys tys')
-                                          }
-                                  , modScope'
-                                  , env')
+                  case runVerify (verifyProgram prog) of
```
Comment (SquidDev) [inline compiler/Main.hs]: I wonder if we should look into shifting this into a monad error or something.

Comment (zardyh) [reply compiler/Main.hs]: We probably should, but that'll be Hard tm

Comment (zardyh) [reply compiler/Main.hs]: because they all have different error types

File: src/Data/Span.hs [hunk 2]
```diff
@@ -63,3 +63,4 @@ instance Semigroup Span where
     | x == internal = y
     | y == internal = x
     | otherwise = error $ "<> spans of different files: " ++ fa ++ ", " ++ fb
+
```
[the remark this answers is not in the document]
Comment (zardyh) [reply src/Data/Span.hs]: damn it sorry

File: compiler/Main.hs (hunk 1 above)
Comment (SquidDev) [reply compiler/Main.hs]: I was thinking something like `` `catchError` (throwError . Verify)`, where `Verify` is some constructor for a union of all error types.

Or we do something with existentials :p.

Comment (zardyh) [reply compiler/Main.hs]: That won't work.

**Comment by zardyh** [comment]
@SquidDev This is done.

**Pull request merged**